common/remote/hooks/git/gitfile.cpp

```diff
@@ -410,10 +413,329 @@ class GitRepositoryDirectoryIterator : implements IDirectoryIterator, public CIn
+            return 1;  // do not recurse - only expand a single level of the directory tree
+        }
+
+        unsigned lenFilename = strlen(filename);
+        unsigned remaining = relDir.length() - matchedPath;
+        if (lenFilename <= remaining)
+        {
+            const char * next = relDir.str() + matchedPath;
+            if (strnicmp(next, filename, lenFilename) == 0)
+            {
+                if (lenFilename == remaining)
+                {
+                    sizes.append(isDirectory ? (offset_t) -1 : 0);
+                    files.append(filename);
+                    oids.emplace_back(*git_tree_entry_id(entry));
+                    return -1;  // found the single match
+                }
+
+                unsigned nextChar = next[lenFilename];
+                if (isPathSepChar(nextChar))
+                {
+                    matchedPath += (lenFilename + 1);
+                    return 0; // recurse
+                }
+
+                // filename only matches a substring of the next directory that needs to match
+            }
+        }
+        return 1;   // skip
+    }
+
+protected:
+    StringAttr gitDirectory;
+    StringAttr revision;
+    StringAttr relDir;
+    StringAttr mask;
+    Owned<IFile> curFile;
+    unsigned curIndex;
+    StringArray files;
+    UInt64Array sizes;
+    std::vector<git_oid> oids;
+    Owned<GitCommitTree> commitTree;
+    bool includeDirs;
+    bool sub;
+    unsigned matchedPath;
+
+    void open()
+    {
+        if (files.isItem(curIndex))
+        {
+            const char *filename = files.item(curIndex);
+            offset_t size = sizes.item(curIndex);
+            const git_oid & oid = oids[curIndex];
+            StringBuffer gitFileName;
+            buildGitFileName(gitFileName, gitDirectory, revision, relDir);
+            // Git ls-tree behaves differently according to whether you put the trailing / on the path you supply.
+            // With /, it gets all files in that directory
+            // Without, it will return just a single match (for the file or dir with that name)
+            // So we are effectively in two different modes according to which we used.
+            char lastChar = gitFileName.charAt(gitFileName.length()-1);
+            // NOTE: / or PATHSEPCHAR - we translated to git representation, but root directory is .git{x}<pathsep>
+            if ((lastChar == '/') || (lastChar == PATHSEPCHAR))
+                gitFileName.append(filename);
+            if (size==(offset_t) -1)
+                curFile.setown(new NewGitRepositoryFile(gitFileName, 0, true, true, commitTree, &oid));
+            else
+                curFile.setown(new NewGitRepositoryFile(gitFileName, size, false, true, commitTree, &oid));
+        }
+        else
+            curFile.clear();
+    }
+};
+
+//--------------------------------------------------------------------------------------------------------------------
+
+static constexpr bool useLibGit = true;
```

Comment:
I assume this will be removed (and the old code removed) once new is proved to work?